decoder/stats.go

```diff
@@ -365,6 +412,156 @@ func updatePokemonStats(old *Pokemon, new *Pokemon, areas []geo.AreaName) {
 	}
 }
 
+func updateRaidStats(old *Gym, new *Gym, areas []geo.AreaName) {
+	if len(areas) == 0 {
+		areas = []geo.AreaName{
+			{
+				Parent: "world",
+				Name:   "world",
+			},
+		}
+	}
+
+	locked := false
+
+	// Loop though all areas
+	for i := 0; i < len(areas); i++ {
+		area := areas[i]
+
+		// Check if RaidSpawnTimestamp has changed, then we can assume raid has disappeared or new raid started.
+		if old == nil || old.RaidSpawnTimestamp != new.RaidSpawnTimestamp {
+
+			if !locked {
+				raidStatsLock.Lock()
+				locked = true
+			}
+
+			countStats := raidCount[area]
+			if countStats == nil {
+				countStats = make(map[int64]areaRaidCountDetail)
+				raidCount[area] = countStats
+			}
+
+			// Raid has started or is active.
+			if new.RaidPokemonId.Int64 != 0 {
+				var countRaids = raidCount[area][new.RaidLevel.Int64]
+				countRaids.count[new.RaidPokemonId.Int64]++
+				raidCount[area][new.RaidLevel.Int64] = countRaids
+			}
+		}
+	}
+
+	if locked {
+		raidStatsLock.Unlock()
+	}
+}
+
+func updateIncidentStats(old *Incident, new *Incident, areas []geo.AreaName) {
+	if len(areas) == 0 {
+		areas = []geo.AreaName{
+			{
+				Parent: "world",
+				Name:   "world",
+			},
+		}
+	}
+
+	locked := false
+
+	// Loop though all areas
+	for i := 0; i < len(areas); i++ {
+		area := areas[i]
+
+		// Check if StartTime has changed, then we can assume a new Incident has appeared.
+		if old == nil || old.StartTime != new.StartTime {
+
+			if !locked {
+				incidentStatsLock.Lock()
+				locked = true
+			}
+
+			invasionStats := invasionCount[area]
+			if invasionStats == nil {
+				invasionStats = &areaInvasionCountDetail{}
+				invasionCount[area] = invasionStats
+			}
+
+			// Exclude Kecleon, Showcases and other UNSET characters for invasionStats.
+			if new.Character != 0 {
+				invasionStats.count[new.Character]++
```

Comment:
so this stats is about rocket invasions and we don't care about other incidents, like kecleon, showcases ? :)

Reply:
Naming was a bit tricky for when it was Incident related and when invasion started. So could have mixed it up somewhere.

I tried to keep it open at least for adding the posibility to have something like Showcase, Kecleon or future Incidents stats. Just purely focused on invasion for now, which is why Kecleon, Showcases and UNSET is excluded from invasionStats.